source/mqtt311_listener.c

```diff
@@ -0,0 +1,291 @@
+        aws_mem_release(entry->allocator, entry);
+    }
+}
+
+static struct aws_mqtt311_callback_set_entry *s_new_311_callback_set_entry(
+    struct aws_mqtt311_callback_set_manager *manager,
+    struct aws_mqtt311_callback_set *callback_set) {
+    struct aws_mqtt311_callback_set_entry *entry =
+        aws_mem_calloc(manager->allocator, 1, sizeof(struct aws_mqtt311_callback_set_entry));
+
+    entry->allocator = manager->allocator;
+    entry->id = manager->next_callback_set_entry_id++;
+    entry->callbacks = *callback_set;
+
+    AWS_LOGF_INFO(
+        AWS_LS_MQTT_GENERAL,
+        "id=%p: MQTT311 callback manager created new entry :%" PRIu64,
+        (void *)manager->connection,
+        entry->id);
+
+    return entry;
+}
+
+uint64_t aws_mqtt311_callback_set_manager_push_front(
+    struct aws_mqtt311_callback_set_manager *manager,
+    struct aws_mqtt311_callback_set *callback_set) {
+
+    AWS_FATAL_ASSERT(
+        aws_event_loop_thread_is_callers_thread(s_mqtt_client_connection_get_event_loop(manager->connection)));
+
+    struct aws_mqtt311_callback_set_entry *entry = s_new_311_callback_set_entry(manager, callback_set);
+
+    aws_linked_list_push_front(&manager->callback_set_entries, &entry->node);
+
+    return entry->id;
+}
+
+void aws_mqtt311_callback_set_manager_remove(
+    struct aws_mqtt311_callback_set_manager *manager,
+    uint64_t callback_set_id) {
+
+    AWS_FATAL_ASSERT(
+        aws_event_loop_thread_is_callers_thread(s_mqtt_client_connection_get_event_loop(manager->connection)));
+
+    struct aws_linked_list_node *node = aws_linked_list_begin(&manager->callback_set_entries);
+    while (node != aws_linked_list_end(&manager->callback_set_entries)) {
+        struct aws_mqtt311_callback_set_entry *entry =
+            AWS_CONTAINER_OF(node, struct aws_mqtt311_callback_set_entry, node);
+        node = aws_linked_list_next(node);
+
+        if (entry->id == callback_set_id) {
+            aws_linked_list_remove(&entry->node);
+
+            AWS_LOGF_INFO(
+                AWS_LS_MQTT_GENERAL,
+                "id=%p: MQTT311 callback manager removed entry id=%" PRIu64,
+                (void *)manager->connection,
+                entry->id);
+            aws_mem_release(entry->allocator, entry);
+            return;
+        }
+    }
+    AWS_LOGF_INFO(
+        AWS_LS_MQTT_GENERAL,
+        "id=%p: MQTT311 callback manager failed to remove entry id=%" PRIu64 ", callback set id not found.",
+        (void *)manager->connection,
+        callback_set_id);
+}
+
+void aws_mqtt311_callback_set_manager_on_publish_received(
+    struct aws_mqtt311_callback_set_manager *manager,
+    const struct aws_byte_cursor *topic,
+    const struct aws_byte_cursor *payload,
+    bool dup,
+    enum aws_mqtt_qos qos,
+    bool retain) {
+
+    struct aws_mqtt_client_connection_311_impl *connection_impl = manager->connection->impl;
+    AWS_FATAL_ASSERT(aws_event_loop_thread_is_callers_thread(connection_impl->loop));
```

Comment:
`s_mqtt_client_connection_get_event_loop` isn't used here and in the below function. I'm not positive it needs to be used here but I was also not 100% it needed to be used above so calling it out just in case it was just missed here.

Reply:
These functions need the downcasted impl anyways, so the loop can be pulled off easily after that.